File: src/test/java/dev/akif/cookbook/api/RecipeControllerTests.java
package dev.akif.cookbook.api;

import dev.akif.cookbook.ingredient.CreateMeasuredIngredient;
import dev.akif.cookbook.ingredient.Ingredient;
import dev.akif.cookbook.ingredient.MeasuredIngredient;
import dev.akif.cookbook.ingredient.Unit;
import dev.akif.cookbook.recipe.Recipe;
import dev.akif.cookbook.recipe.RecipeService;
import dev.akif.cookbook.recipe.SearchParameters;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.MediaType;
import org.springframework.test.web.reactive.server.WebTestClient;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;

@WebFluxTest
class RecipeControllerTests {
    @Autowired
    private WebTestClient testClient;

    @MockBean
    private RecipeService recipeService;

    private final Recipe pancakesRecipe = new Recipe(
            1L,
            "Pancakes",
            4,
            Map.of(
                    "For Batter", List.of(
                            new MeasuredIngredient(new Ingredient(1, "Egg", true), 2, Unit.QUANTITY),
                            new MeasuredIngredient(new Ingredient(2, "Sugar", true), 0.5, Unit.CUPS),
                            new MeasuredIngredient(new Ingredient(3, "Flour", true), 2, Unit.CUPS),
                            new MeasuredIngredient(new Ingredient(4, "Milk", true), 1.5, Unit.CUPS),
                            new MeasuredIngredient(new Ingredient(5, "Oil", true), 2, Unit.SPOONS),
                            new MeasuredIngredient(new Ingredient(6, "Baking Powder", true), 1, Unit.PACKS),
                            new MeasuredIngredient(new Ingredient(7, "Vanilla", true), 1, Unit.PACKS),
                            new MeasuredIngredient(new Ingredient(8, "Salt", true), 1, Unit.PINCHES)
                    ),
                    "For Cooking", List.of(
                            new MeasuredIngredient(new Ingredient(5, "Oil", true), 2, Unit.SPOONS)
                    ),
                    "For Serving", List.of(
                            new MeasuredIngredient(new Ingredient(9, "Honey", true), 1, Unit.SPOONS)
                    )
            ),
            List.of(
                    "Break eggs and add sugar.",
                    "Mix until creamy.",
                    "Add oil and milk and continue mixing.",
                    "Add flour, baking powder, vanilla and salt.",
                    "Mix until smooth.",
                    "Slightly oil a pan and heat. Reduce heat once pan is hot.",
                    "Pour mixture into pan.",
                    "Cook until bubbles appear on top and flip to cook the other side.",
                    "Cook all of the batter the same way.",
                    "Pour honey on top and serve."
            )
    );

    private final RecipeDTO pancakesRecipeDTO = new RecipeDTO(
            1L,
            "Pancakes",
            4,
            Map.of(
                    "For Batter", List.of(
                            new MeasuredIngredientDTO("Egg", true, 2, Unit.QUANTITY),
                            new MeasuredIngredientDTO("Sugar", true, 0.5, Unit.CUPS),
                            new MeasuredIngredientDTO("Flour", true, 2, Unit.CUPS),
                            new MeasuredIngredientDTO("Milk", true, 1.5, Unit.CUPS),
                            new MeasuredIngredientDTO("Oil", true, 2, Unit.SPOONS),
                            new MeasuredIngredientDTO("Baking Powder", true, 1, Unit.PACKS),
                            new MeasuredIngredientDTO("Vanilla", true, 1, Unit.PACKS),
                            new MeasuredIngredientDTO("Salt", true, 1, Unit.PINCHES)
                    ),
                    "For Cooking", List.of(
                            new MeasuredIngredientDTO("Oil", true, 2, Unit.SPOONS)
                    ),
                    "For Serving", List.of(
                            new MeasuredIngredientDTO("Honey", true, 1, Unit.SPOONS)
                    )
            ),
            List.of(
                    "Break eggs and add sugar.",
                    "Mix until creamy.",
                    "Add oil and milk and continue mixing.",
                    "Add flour, baking powder, vanilla and salt.",
                    "Mix until smooth.",
                    "Slightly oil a pan and heat. Reduce heat once pan is hot.",
                    "Pour mixture into pan.",
                    "Cook until bubbles appear on top and flip to cook the other side.",
                    "Cook all of the batter the same way.",
                    "Pour honey on top and serve."
            )
    );

    private final CreateRecipeDTO createPancakesDTO = new CreateRecipeDTO(
            "Pancakes",
            4,
            Map.of(
                    "For Batter", List.of(
                            new MeasuredIngredientDTO("Egg", true, 2, Unit.QUANTITY),
                            new MeasuredIngredientDTO("Sugar", true, 0.5, Unit.CUPS),
                            new MeasuredIngredientDTO("Flour", true, 2, Unit.CUPS),
                            new MeasuredIngredientDTO("Milk", true, 1.5, Unit.CUPS),
                            new MeasuredIngredientDTO("Oil", true, 2, Unit.SPOONS),
                            new MeasuredIngredientDTO("Baking Powder", true, 1, Unit.PACKS),
                            new MeasuredIngredientDTO("Vanilla", true, 1, Unit.PACKS),
                            new MeasuredIngredientDTO("Salt", true, 1, Unit.PINCHES)
                    ),
                    "For Cooking", List.of(
                            new MeasuredIngredientDTO("Oil", true, 2, Unit.SPOONS)
                    ),
                    "For Serving", List.of(
                            new MeasuredIngredientDTO("Honey", true, 1, Unit.SPOONS)
                    )
            ),
            List.of(
                    "Break eggs and add sugar.",
                    "Mix until creamy.",
                    "Add oil and milk and continue mixing.",
                    "Add flour, baking powder, vanilla and salt.",
                    "Mix until smooth.",
                    "Slightly oil a pan and heat. Reduce heat once pan is hot.",
                    "Pour mixture into pan.",
                    "Cook until bubbles appear on top and flip to cook the other side.",
                    "Cook all of the batter the same way.",
                    "Pour honey on top and serve."
            )
    );

    @Test
    void gettingRecipes() {
        Mockito.when(
                recipeService.getAll(new SearchParameters(null, null, null, null, null, null))
        ).thenReturn(Flux.just(pancakesRecipe));

        testClient.get()
                .uri("/recipes")
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus().isOk()
                .expectBody(new ParameterizedTypeReference<List<RecipeDTO>>() {
                })
                .value(recipes -> {
                    assertEquals(1, recipes.size());
                    assertEquals(pancakesRecipeDTO, recipes.get(0));
                });
    }

    @Test
    void creatingRecipe() {
        Mockito.when(
                recipeService.create(
                        createPancakesDTO.name(),
                        createPancakesDTO.numberOfServings(),
                        Map.of(
                                "For Batter", List.of(
                                        new CreateMeasuredIngredient("Egg", true, 2, Unit.QUANTITY),
                                        new CreateMeasuredIngredient("Sugar", true, 0.5, Unit.CUPS),
                                        new CreateMeasuredIngredient("Flour", true, 2, Unit.CUPS),
                                        new CreateMeasuredIngredient("Milk", true, 1.5, Unit.CUPS),
                                        new CreateMeasuredIngredient("Oil", true, 2, Unit.SPOONS),
                                        new CreateMeasuredIngredient("Baking Powder", true, 1, Unit.PACKS),
                                        new CreateMeasuredIngredient("Vanilla", true, 1, Unit.PACKS),
                                        new CreateMeasuredIngredient("Salt", true, 1, Unit.PINCHES)
                                ),
                                "For Cooking", List.of(
                                        new CreateMeasuredIngredient("Oil", true, 2, Unit.SPOONS)
                                ),
                                "For Serving", List.of(
                                        new CreateMeasuredIngredient("Honey", true, 1, Unit.SPOONS)
                                )
                        ),
                        createPancakesDTO.instructions()
                )
        ).thenReturn(Mono.just(pancakesRecipe));

        testClient.post()
                .uri("/recipes")
                .body(Mono.just(createPancakesDTO), CreateRecipeDTO.class)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus().isCreated()
                .expectBody(RecipeDTO.class)
                .value(recipe -> {
                    assertEquals(pancakesRecipeDTO, recipe);
                });
    }
}

File: src/test/kotlin/dev/akif/cookbook/api/RecipeControllerTests.kt
package dev.akif.cookbook.api

import dev.akif.cookbook.*
import dev.akif.cookbook.ingredient.IngredientRepository
import dev.akif.cookbook.recipe.RecipeRepository
import dev.akif.cookbook.recipe.ingredient.RecipeIngredientRepository
import dev.akif.cookbook.recipe.instruction.RecipeInstructionRepository
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.MethodOrderer
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestMethodOrder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.reactive.AutoConfigureWebTestClient
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.MediaType
import org.springframework.test.context.DynamicPropertyRegistry
import org.springframework.test.context.DynamicPropertySource
import org.springframework.test.web.reactive.server.WebTestClient
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers
import reactor.core.publisher.Mono

@ExperimentalCoroutinesApi
@AutoConfigureWebTestClient
@SpringBootTest
@Testcontainers
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
class RecipeControllerTests(
    @Autowired private val testClient: WebTestClient,
    @Autowired private val recipes: RecipeRepository,
    @Autowired private val ingredients: IngredientRepository,
    @Autowired private val recipeIngredients: RecipeIngredientRepository,
    @Autowired private val recipeInstructions: RecipeInstructionRepository
) {
    companion object {
        @Container
        val postgresql = PostgreSQLContainer<Nothing>("postgres:15.1-alpine")

        @DynamicPropertySource
        fun properties(registry: DynamicPropertyRegistry) {
            registry.add("spring.flyway.url") { postgresql.jdbcUrl }
            registry.add("spring.r2dbc.url") { "r2dbc:postgresql://${postgresql.host}:${postgresql.getMappedPort(PostgreSQLContainer.POSTGRESQL_PORT)}/${postgresql.databaseName}" }
            registry.add("spring.r2dbc.username") { postgresql.username }
            registry.add("spring.r2dbc.password") { postgresql.password }
        }
    }

    @Order(1)
    @Test
    fun gettingRecipes() = runTest {
        testClient.get()
            .uri("/recipes")
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus().isOk
            .expectBody(object : ParameterizedTypeReference<List<RecipeDTO>>() {})
            .value { recipes: List<RecipeDTO> ->
                assertEquals(1, recipes.size)
                assertEquals(pancakesRecipeDTO, recipes[0].copy(id = pancakesRecipeDTO.id))
            }
    }

    @Order(2)
    @Test
    fun creatingRecipe() = runTest {
        recipeInstructions.deleteAll()
        recipeIngredients.deleteAll()
        ingredients.deleteAll()
        recipes.deleteAll()

        testClient.post()
            .uri("/recipes")
            .body(Mono.just(createPancakesDTO), CreateRecipeDTO::class.java)
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus().isCreated
            .expectBody(RecipeDTO::class.java)
            .value { recipe: RecipeDTO -> assertEquals(pancakesRecipeDTO, recipe.copy(id = pancakesRecipeDTO.id)) }
    }

    @Order(3)
    @Test
    fun updatingRecipe() = runTest {
        val recipeId = recipes.findAllBy(null, null).first().id!!

        testClient.put()
            .uri("/recipes/$recipeId")
            .body(Mono.just(updatePancakesDTO), UpdateRecipeDTO::class.java)
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus().isOk
            .expectBody(RecipeDTO::class.java)
            .value { recipe: RecipeDTO -> assertEquals(updatedPancakesRecipeDTO.copy(id = recipeId), recipe) }
    }
}
